### DISCUSSION TITLE
Handling an enum with an interior from external library

### DISCUSSION THREAD

**Discussion by AndrewJSchoen** [PyO3/pyo3#2872] [Questions]
Hi all,

I am trying to make bindings for an enum that contains values from an external library, which itself unfortunately doesn't currently support PyO3 (nalgebra). Currently, I have the following enum:

```rust
#[repr(C)]
#[derive(Serialize,Deserialize,Clone,Debug)]
#[cfg_attr(feature = "pybindings", derive(FromPyObject))]
pub enum Goal {
    Translation(Translation3<f64>), // From nalgebra
    Rotation(UnitQuaternion<f64>), // From nalgebra
    Scalar(f64),
    Size(Vector3<f64>), // From nalgebra
    Ellipse(Ellipse), // My own struct
    RotationRange(RotationRange),  // My own struct
    ScalarRange(ScalarRange) // My own struct
}

#[cfg(feature = "pybindings")]
impl IntoPy<PyObject> for Goal {
    fn into_py(self, py: Python) -> PyObject {
        match self {
            Self::Translation(obj) => PyTranslation{value:obj}.into_py(py), // My own struct wraps the nalgebra version and converts that
            Self::Rotation(obj) => PyRotation{value:obj}.into_py(py),  // My own struct wraps the nalgebra version and converts that
            Self::Scalar(obj) => obj.into_py(py),
            Self::Size(obj) => PySize{value:obj}.into_py(py),  // My own struct wraps the nalgebra version and converts that
            Self::Ellipse(obj) => obj.into_py(py),
            Self::RotationRange(obj) => obj.into_py(py),
            Self::ScalarRange(obj) => obj.into_py(py),
        }
    }
}

// Example implementation for PyTranslation wrapper:
#[cfg_attr(feature = "pybindings", pyclass(name="Translation"))]
#[derive(Clone,Debug)]
pub struct PyTranslation {
    pub value: Translation3<f64>
}
```

As you can see, I created wrappers (`PyTranslation`, `PyRotation`, `PySize`) that do have PyO3 bindings. However, I still get an error that 

```
the trait `PyClass` is not implemented for `Translation<f64, 3_usize>`
```
(That's the Translation3 from the external I wrap). Any suggestions on how to handle this? I also have WASM bindings where this all works, so ideally I don't want to break all those by wrapping these 3 external structs in something all the time.

Thanks for the help!

**Comment by adamreichold** [comment]
I think the error messages originate from

```rust
#[cfg_attr(feature = "pybindings", derive(FromPyObject))]
```

which tries to apply `FromPyObject::extract` to the enum's inner values. I think you will need a similar manual implement for `FromPyObject` as you have for `IntoPy`.

(Besides that, it might be of interest to you that the next version of the `numpy` crate will contain some integration with `nalgebra`, c.f. https://github.com/PyO3/rust-numpy/pull/347 However, I fear that is still not sufficient for your use case but maybe you want to collaborate on that over at the rust-numpy project.)

**Reply by AndrewJSchoen** [reply]
That makes sense conceptually. Are there any examples of how to implement this from scratch? I couldn't find any in the documentation.

**Reply by adamreichold** [reply]
Not sure, but the main problem is probably recovering your `PyTranslation` etc. wrappers, i.e. extract a `PyTranslation` and then recover the inner `Translation3<f64>`, e.g.

```rust
impl<'source> FromPyObject<'source> for Goal {
  fn extract(ob: &'source PyAny) -> PyResult<Self> {
    if let Ok(ob) = PyTranslation::extract(ob) {
      return Ok(Self::Translation(ob.value));
    }

    if let Ok(ob) = PyRotation::extract(ob) {
      return Ok(Self::Rotation(ob.value));
    }

    ...
  }
}
```

**Reply by adamreichold** [reply]
(Generally speaking, the `FromPyObject` implemenation of an enum will usually compose existing `FromPyObject` implementations as alternatives.)

**Reply by AndrewJSchoen** [reply]
Got it, that makes sense. I think what I was doing was _technically_ correct, but probably much more wasteful in terms of computation (checking attributes). Thanks!
